### ISSUE TITLE
tic80 doesn't find my required lua files

### ISSUE THREAD

**Issue by atesin** [nesbox/TIC-80#2426]
hi, i am in the programming developing/learning stage

to keep some order while in developing/testing i pretend to separate the program in various lua files, and call them with `require('my-lib')` in my cart code... previously i created a subfolder inside tic80 carts folder, to keep order i pretend later to create a separate folder for each project, with cart and all needed files inside

currently my folder structure looks like

![imagen](https://github.com/nesbox/TIC-80/assets/8086106/348a1d70-9305-4a46-83dd-7fa2bfb6cfca)

in tic80 console i must type these commands

![imagen](https://github.com/nesbox/TIC-80/assets/8086106/25dd18cd-862e-485d-b3c1-90b3d2ec7b8f)

ahd the code looks like

![imagen](https://github.com/nesbox/TIC-80/assets/8086106/878acba5-2d97-4064-a74d-80ca6264e364)

but when i try to run the cart, it crashes and these messages appear in console

![imagen](https://github.com/nesbox/TIC-80/assets/8086106/4c40d77b-5630-44f5-9542-a1be4f840a50)

... i can clearly see the line `no file '.\oop.lua'`, but i swear this file exists as `.\` stands for "current directory", and current directory inside tic80 is `<cart folder>\001-oop` as i recently cd'ed and console prompt shows

i think it must be a bug, that a required file exists in current directory, and current directory being expanded to tic80 console's current directory, by default, as console prompt shows  .... but i see currently is not the case :(

**Comment by joshgoebel** [comment]
`require` simple isn't supported. All your code has to be placed in the single main file.

**Comment by TimotheeGreg** [comment]
> `require` simple isn't supported. All your code has to be placed in the single main file.

I may misunderstand, but I think it does, see #2352

**Comment by TimotheeGreg** [comment]
I never used it, not sure I can help but I try anyway:

Did you look at https://github.com/nesbox/TIC-80/wiki/Using-require-to-load-external-code-into-your-cart ? From what I understand the libraries should be in `TIC-80/lua/module.lua` file.

Also, I think current directory in TIC-80 is always `TIC-80` and not where your cart is or where you `cd` #775 

However, note that the use of `require` could be disabled in future update: #2352

**Comment by atesin** [comment]
> Did you look at https://github.com/nesbox/TIC-80/wiki/Using-require-to-load-external-code-into-your-cart ? From what I understand the libraries should be in `TIC-80/lua/module.lua` file.

yes i tried what that wiki page says and doesn't work (maybe worked in a previous tic80 version)... and for regular lua files other than modules or libraries (e.g. grouplng similar functions, etc.) should work in tic80 command console "cwd" (current working directory, shown in prompt synbol), or the dir where current cart were loaded 

> Also, I think current directory in TIC-80 is always `TIC-80` and not where your cart is or where you `cd` #775

i agree with @moonlit in that this behavior is confusional, prone to provoke errors and should be fixed

> However, note that the use of `require` could be disabled in future update: #2352

`require` lua function could be very useful, i think it should be fixed in its security issues (the right way) instead being disabled (the fast and lazy way)... check my comment there for more details: https://github.com/nesbox/TIC-80/issues/2352#issuecomment-1871701386

**Comment by atesin** [comment]
i am using this dirty workaround:

- create the subdirectory `<tic80-install>/project-name`
- place all lua files here, ex.: `project-main.lua`, `functions.lua`
  - `project-main.lua` is the file that has header comments, `TIC()` function, and others
- create the cartridge file `<tic80-carts>/project-cart.tic` with just this content
```
require('project-name.functions') -- will map to "<tic80-install>/project-name/functions.lua"
-- require('project-name.more-files')
require('project-name.project-main') -- main file at LAST because include also execute files
```
- edit files in `<tic80-install>/project-name` with zerobrane or another external editor
- when pass to production, concatenate all files and run in tic console `import code joined-files.lua` and save

**Comment by joshgoebel** [comment]
> not true, i tried with some of dirs shown in error message and they worked..

There is a huge difference between "it works" and "it is supported".  I'll repeat: This is not supported. Some people wish it was (there is a ticket for this), but it's currently not.  And there is also a ticket to "break" it fully as it's a security risk - and not supported.

**Comment by atesin** [comment]
@joshgoebel ok

**Comment by borbware** [comment]
the behaviour indeed can be a bit erratic: for *some* computers, i've had to add the following line to my `main.lua` for require to work:

`package.path = package.path..";C:/path/to/your/directory/?.lua"`

It adds your working directory to Lua's package.path so Lua can find your files.

**Comment by atesin** [comment]
@borbware : as i thought `config.cart` is loaded frist any cartridge, i tried to set this permanently by typing `config` command and adding this line in code editor:
```
package.path='%APPDATA%/com.nesbox.tic/TIC-80/?.lua;%APPDATA%/com.nesbox.tic/TIC-80/?/init.lua'..package.path

THEME=
{
	CODE =
	{
...
...
```
but didn't worked :(  ...   is there any way to do this permanently? (i think it should come from stock)

**Comment by atesin** [comment]
i found a better workaround that works:

as a response to my previous response to @TimotheeGreg , i put together these 2 facts:
- edit start folder in tic80 shortcut to `%APPDATA%/com.nesbox.tic/TIC-80`
- traceback show package.path with `.\?.lua` and `.\?\init.lua`

when i edit tic80.exe shortcut start folder and add `require('req-test')` to code editor, it works for `(carts_folder)/req-test.lua` and `(carts_folder)/req-test/init.lua` =D

**Comment by ephetic** [comment]
TIC-80 should really support some form of multi-file support.  It need not be the native `require` function, but I think should be something that would allow people to easily refactor their `require` hacks to use the new solution for their previous sandbox-friendly needs.  

- `require` -like syntax
- `edit <otherfile>` support, or similar
- automatically bundle/concat files on export 

I don't think this violates the retro spirit of TIC-80 more than the editor outline and bookmarks views do already and I'm not sure what use the console (e.g. mkdir) is without this sort of feature.

**Comment by scambier** [comment]
> TIC-80 should really support some form of multi-file support. It need not be the native `require` function, but I think should be something that would allow people to easily refactor their `require` hacks to use the new solution for their previous sandbox-friendly needs.

You can already use multiple files: https://github.com/nesbox/TIC-80/wiki/tools#bundling

I hear you though, it would be nice if TIC-80 supported that directly, but:
- There are many languages supported by TIC
- They all manage modules differently
- They already all require external tools for bundling

That would IMO greatly bloat TIC-80, for a problem that is already mostly solved by one of the tools listed in the wiki.

Granted, they only work with the PRO version (to my knowledge), but I think it's fine. The fact that `require()` works with non-pro is a quirk that only works with Lua.
